### PR TITLE
upgrade workflow design

### PR DESCRIPTION
**Pull request by leejanee**
upgrade [workflow design](https://github.com/oam-dev/kubevela/blob/master/design/vela-core/workflow_policy.md)

### CODE REVIEW

File: design/vela-core/workflow_policy.md [hunk 1]
```diff
@@ -199,38 +201,112 @@ Here are the steps in Workflow Manager:
...
 
-- If `continue.return == true`, Task Manager will update the `observedAppGeneration` for this step in status.
+
+### Workflow Operation
+
+#### Terminate Workflow
+
+If the execution of the workflow does not meet expectations, it may be necessary to terminate the workflow
+
+There are two ways to achieve that
```
Comment (hongchaodeng) [inline design/vela-core/workflow_policy.md:241]:
```suggestion
There are two ways to achieve that:
```

Comment (hongchaodeng) [reply design/vela-core/workflow_policy.md:241]: The following two approaches are not mutual-exclusive. The op.Break can be captured and then report terminated status.

File: design/vela-core/workflow_policy.md [hunk 2]
```diff
@@ -199,38 +201,112 @@ Here are the steps in Workflow Manager:
...
+```yaml
+kind: Application
+spec:
+  componnets: ...
+workflow:
+  steps:
+  - name: manual-approve
+    type: suspend   
+ ```
+  The workflow automatically pauses when the suspend-type task is started
```
Comment (hongchaodeng) [inline design/vela-core/workflow_policy.md:280]: How to resume?

File: design/vela-core/workflow_policy.md [hunk 3]
```diff
@@ -199,38 +201,112 @@ Here are the steps in Workflow Manager:
...
+    name: foo
+  status:
+    phase: runningWorkflow
+    workflow:
+      stepIndex: 1
+      terminated: true
+      steps:
+      - name: ...
+ ```
+#### Pause & Resume Workflow
```
Comment (hongchaodeng) [inline design/vela-core/workflow_policy.md:267]: This section is confusing by showing two points directly. Please add some description about why you make two points.

File: design/vela-core/workflow_policy.md [hunk 4]
```diff
@@ -199,38 +201,112 @@ Here are the steps in Workflow Manager:
...
+    phase: runningWorkflow
+    workflow:
+      stepIndex: 1
+      suspend: true
+      steps:
+      - name: ... 
+ ```
+#### Restart Workflow
+
+The workflow will be executed in the following two cases
```
Comment (hongchaodeng) [inline design/vela-core/workflow_policy.md:298]:
```suggestion
The workflow will be restarted in the following two cases:
```

File: design/vela-core/workflow_policy.md [hunk 5]
```diff
@@ -199,38 +201,112 @@ Here are the steps in Workflow Manager:
...
+1. The application spec changes
```
Comment (hongchaodeng) [inline design/vela-core/workflow_policy.md:299]: Please describe what happens to the current state during restart

File: design/vela-core/workflow_policy.md [hunk 6]
```diff
@@ -199,38 +201,112 @@ Here are the steps in Workflow Manager:
...
+2. Modify the value of the `status.phase` field to "runningWorkflow" and clear the status of the workflow
+
+```yaml
+  kind: Application
+  metadata:
+    name: foo
+  status:
+    phase: runningWorkflow
+    workflow:
```
Comment (hongchaodeng) [inline design/vela-core/workflow_policy.md:308]: `workflow: {}`

**Review by hongchaodeng — COMMENTED** [review]
Besides existing comments, there are a few more issues I think we should resolve:

- The mechanism of how CUE task is executed, how to report status is not concrete.
- The limitation of reentrant tasks.
- The limitation of synchronous execution of CUE tasks is not discussed.
- The diagram is confusing:
  - What does 'Create Task' in the middle mean? => call TaskManager
  - Why are there two 'Execute CUE tasks' rectangles? => The second one should be changed to 'parse do steps'.
  - If continue.return is true, it should run next task, right?
  - How to run next task when return is true?
- Need to clarify how to execute all steps in order in a CUE Task including checking the 'do' field.
- In diagram, 'for each workflow step' should be changed to 'get current stepIndex', also need to increment stepIndex at the end.

**Comment by leejanee** [comment]
@hongchaodeng please review.

**Review by hongchaodeng — APPROVED** [review]
lgtm

**Pull request merged**